pkg/controller/generic/transform/controller.go

```diff
@@ -196,6 +196,10 @@ func (ctrl *Controller[Input, Output]) Run(ctx context.Context, r controller.Run
 	for {
 		select {
 		case <-ctx.Done():
+			if ctrl.options.onShutdownCallback != nil {
```

Comment:
this should better be a `defer`

Reply:
I don't think that discussion has anything about _successful_ shutdown. On any exit you want to make sure that all goroutines/workers terminated.

Not only that, you have to use a context which is bound to the controller execution, so if the controller aborts on any reason, context is canceled.

Reply:
Maybe we can do something like this:
```go
ctx, cancel := context.WithCancel(ctx)
defer cancel()

stop := context.AfterFunc(ctx, func() {
	if ctrl.options.onShutdownCallback != nil {
		ctrl.options.onShutdownCallback()
	}
})
defer stop()
```

?

Reply:
hmm... can't we just `defer ...onShutdownCallback()` before `defer cancel()` to make sure it gets called after the context is canceled?

Reply:
I think it would work fine ⬆️ 

Wdyt about allowing to return error from this callback overall? Should I ditch it or keep (despite being in defer)?

Reply:
I feel such callbacks might get a logger as an argument, but reporting an error feels like too late (as the controller is already terminating either as a result of error or as the runtime is terminating).

also I wonder if we could use `goleak` in the unit-tests as a defence in depth: make sure controller terminates and cleans up all goroutines

Reply:
(I mean `goleak` in Omni, runtime tests here should be already using it)

Reply:
@smira @DmitriyMV I (think I) did the changes. Could you guys have another look to it?